subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -168,27 +164,47 @@ private CompleteFileSystemLocationSnapshot snapshot(String location, Snapshottin
     }
 
     private CompleteFileSystemLocationSnapshot readLocation(String location) {
-        return virtualFileSystem.getRoot().getSnapshot(location)
+        return readSnapshotFromLocation(location, () -> snapshot(location, SnapshottingFilter.EMPTY));
+    }
+
+    private CompleteFileSystemLocationSnapshot readSnapshotFromLocation(
+        String location,
+        Supplier<CompleteFileSystemLocationSnapshot> readFromDisk
+    ) {
+        return readSnapshotFromLocation(
+            location,
+            Function.identity(),
+            readFromDisk
+        );
+    }
+
+    private <T> T readSnapshotFromLocation(
+        String location,
+        Function<CompleteFileSystemLocationSnapshot, T> snapshotProcessor,
+        Supplier<T> readFromDisk
+    ) {
+        return virtualFileSystem.getSnapshot(location)
+            .map(snapshotProcessor)
+            // Avoid snapshotting the same location at the same time
             .orElseGet(() -> producingSnapshots.guardByKey(location,
-                () -> virtualFileSystem.getRoot().getSnapshot(location).orElseGet(() -> snapshot(location, SnapshottingFilter.EMPTY)))
-            );
+                () -> virtualFileSystem.getSnapshot(location)
+                    .map(snapshotProcessor)
+                    .orElseGet(readFromDisk)
+            ));
     }
 
     @Override
     public void write(Iterable<String> locations, Runnable action) {
         writeListener.locationsWritten(locations);
-        virtualFileSystem.update((outerRoot, outerDiffListener) -> {
-            for (String location : locations) {
-                virtualFileSystem.update((innerRoot, diffListener) -> innerRoot.invalidate(location, diffListener));
-            }
-            return virtualFileSystem.getRoot();
-        });
+        for (String location : locations) {
+            virtualFileSystem.invalidate(location);
```

Comment:
I think `invalidate()` should take all the locations and lock only once.

Reply:
Done.